src/frob.cc

```diff
@@ -90,6 +89,96 @@ unsigned long ByteData(unsigned long addr) {
+  // TODO: This function only works for Python >= 3.3. Is it also possible to
+  // support older versions of Python 3?
+
+  // TODO: Can we guarantee that the same padding is used for the bitfield?
+  const std::unique_ptr<uint8_t[]> unicode_bytes =
+      PtracePeekBytes(pid, addr, sizeof(PyASCIIObject));
+  PyASCIIObject *unicode =
+      reinterpret_cast<PyASCIIObject *>(unicode_bytes.get());
+
+  // Because both the filename and function name string objects are made by the
+  // Python interpreter itself, we can probably assume they are compact. This
+  // means that the data immediately follows the object, and is of type {ASCII,
+  // Latin-1, UCS-2, UCS-4}.
+  assert(unicode->state.compact);
+
+  const long str_offset = unicode->state.ascii ? sizeof(PyASCIIObject)
+                                               : sizeof(PyCompactUnicodeObject);
+
+  // NOTE: From CPython commit c47adb04 onwards the kind matches directly to
+  // character size. This is different from the unicode format specification
+  // outlined in PEP 393, which still had only two bits allocated to the kind
+  // field.
+  const long ch_size = unicode->state.kind;
+  const long str_length = ch_size * unicode->length;
+  const std::unique_ptr<uint8_t[]> bytes =
+      PtracePeekBytes(pid, addr + str_offset, str_length);
+
+  std::ostringstream dump;
+
+  for (int i = 0; i < str_length; i += ch_size) {
+    Py_UCS4 ch = 0;
+
+    switch (unicode->state.kind) {
+      case PyUnicode_1BYTE_KIND:
+        ch = bytes[i];
+        break;
+      case PyUnicode_2BYTE_KIND: {
+        Py_UCS2 *data_2 = reinterpret_cast<Py_UCS2 *>(&bytes.get()[i]);
+        ch = *data_2;
+        break;
+      }
+      case PyUnicode_4BYTE_KIND: {
+        Py_UCS4 *data_4 = reinterpret_cast<Py_UCS4 *>(&bytes.get()[i]);
+        ch = *data_4;
+        break;
+      }
+      default:
+        // We are not supposed to come here, as the WCHAR kind is not supported
+        // when the object is compact.
+        assert(false);
+        break;
+    }
+
+    // TODO: Is it alright to assume a lack of surrogates. They might be present
+    // in the UCS-2 representation if the UTF-16 approach is used. We currently
+    // assume that CPython will instead use UCS-4 for such characters, instead
+    // of using surrogates.
+
+    // Below section is taken from CPython's STRINGLIB(utf8_encoder) routine.
+    // The differences are that (1) we use a string builder instead of a char
+    // buffer, and (2) that we skip the surrogate handling entirely.
+    if (ch < 0x80) {
+      /* Encode ASCII */
+      dump << (char)ch;
+    } else if (ch < 0x0800) {
+      /* Encode Latin-1 */
+      dump << (char)(0xc0 | (ch >> 6));
+      dump << (char)(0x80 | (ch & 0x3f));
+    } else if (ch < 0x10000) {
+      dump << (char)(0xe0 | (ch >> 12));
+      dump << (char)(0x80 | ((ch >> 6) & 0x3f));
+      dump << (char)(0x80 | (ch & 0x3f));
+    } else /* ch >= 0x10000 */
+    {
```

Comment:
Can you move this brace up a line? You can still keep the inline comment on line 166. If you run the code through `clang-format` it should format things properly, assuming it doesn't completely butcher the bit twiddling stuff here.